### PR TITLE
Expose canonical URLs

### PR DESCRIPTION
**Pull request by prashtx**
Fix an issue preventing the existing canonical-url implementation from kicking in.

Set actual canonical URLs for paths where the mobile-web and desktop URLs don't match.

### CODE REVIEW

**Comment by prashtx** [comment]
👓 @schwers

File: src/routes.jsx [hunk 1]
```diff
@@ -473,7 +475,19 @@ function routes(app) {
...
-    props.data.set('listing', app.api.links.get(listingOpts));
+    const listingPromise = app.api.links.get(listingOpts);
+    props.data.set('listing', listingPromise);
+
+    // For the canonical path, we need the subreddit, listing ID, and listing title.
+    if (ctx.params.subreddit && ctx.params.listingTitle && ctx.params.listingId) {
+      const { subreddit, listingTitle, listingId } = ctx.params;
+      props.canonicalPath = `/r/${subreddit}/comments/${listingId}/${listingTitle}/`;
+    } else {
+      listingPromise.then(listing => {
```
Comment (schwers) [inline src/routes.jsx]: Should this be added to `props.data`? `horse-react/server` will call `Promise.all` on `props.data`[[1]](https://github.com/reddit/node-horse-react/blob/master/src/server.es6.js#L63) before rendering. I'm worried about a race-condition when `listingPromise` resolves last for some reason and `cannonicalPath` isn't set before rendering.

File: src/routes.jsx [hunk 2]
```diff
@@ -752,10 +773,12 @@ function routes(app) {
   }
 
   router.get('saved', '/u/:user/saved', function * () {
+    this.props.canonicalPath = `/user/${this.params.user}/saved/`;
```
Comment (schwers) [inline src/routes.jsx]: Do we know if crawlers visit these pages with some authentication? Saved and Hidden seem like pages we don't want crawled in general.

**Comment by schwers** [comment]
💅  Patch looks good overall, I just have a minor question and an implementation note.

File: src/routes.jsx (hunk 2 above)
Comment (prashtx) [inline src/routes.jsx]: We actually block these via `robots.txt`, but it bugged me that we had canonical tags that were not actually canonical...

**Comment by prashtx** [comment]
💇

File: src/routes.jsx (hunk 2 above)
Comment (schwers) [inline src/routes.jsx]: Ah I see, okay.

**Comment by schwers** [comment]
:fish:

**Comment by prashtx** [comment]
@schwers: caught an issue with lack of trailing `/` in listings canonicals.

💇

**Comment by schwers** [comment]
:fish:

**Pull request merged**